File: backend/app/schemas/websocket.py
from __future__ import annotations

import enum
from typing import Any, Literal

from pydantic import BaseModel, Field


class ClientEventType(str, enum.Enum):
    SESSION_START = "session.start"
    AUDIO_CHUNK = "audio.chunk"
    AUDIO_END = "audio.end"
    CANDIDATE_TEXT = "candidate.text"  # testing shortcut: bypasses STT
    SESSION_END = "session.end"
    PING = "ping"


class ServerEventType(str, enum.Enum):
    SESSION_READY = "session.ready"
    INTERVIEWER_TRANSCRIPT = "interviewer.transcript"
    INTERVIEWER_STATE = "interviewer.state"
    CANDIDATE_TRANSCRIPT = "candidate.transcript"
    INTERVIEWER_RESPONSE = "interviewer.response"
    AUDIO_START = "audio.start"
    AUDIO_CHUNK = "audio.chunk"
    AUDIO_END = "audio.end"
    SILENCE_PROMPT = "silence.prompt"
    ASSESSMENT_STARTED = "assessment.started"
    ASSESSMENT_COMPLETED = "assessment.completed"
    SESSION_COMPLETED = "session.completed"
    ERROR = "error"
    PONG = "pong"


class ClientEvent(BaseModel):
    """A message sent by the client (text frame)."""

    type: ClientEventType
    data: Any | None = Field(default=None, description="Optional JSON payload for the event.")


class ServerEvent(BaseModel):
    """A message sent to the client (text frame)."""

    type: ServerEventType
    data: dict[str, Any] | None = Field(default=None)


def error_event(code: str, message: str) -> ServerEvent:
    return ServerEvent(type=ServerEventType.ERROR, data={"code": code, "message": message})


def audio_start_event(format: str, sample_rate: int) -> ServerEvent:
    return ServerEvent(
        type=ServerEventType.AUDIO_START,
        data={"format": format, "sample_rate": sample_rate},
    )


# -- Event payload builders -------------------------------------------------


def session_ready(session_id: str) -> ServerEvent:
    return ServerEvent(type=ServerEventType.SESSION_READY, data={"session_id": session_id})


def interviewer_transcript(text: str) -> ServerEvent:
    return ServerEvent(type=ServerEventType.INTERVIEWER_TRANSCRIPT, data={"text": text})


def interviewer_state(state: Literal["speaking", "listening", "thinking"]) -> ServerEvent:
    return ServerEvent(type=ServerEventType.INTERVIEWER_STATE, data={"state": state})


def candidate_transcript(text: str) -> ServerEvent:
    return ServerEvent(type=ServerEventType.CANDIDATE_TRANSCRIPT, data={"text": text})


def interviewer_response(text: str, stage: str) -> ServerEvent:
    return ServerEvent(type=ServerEventType.INTERVIEWER_RESPONSE, data={"text": text, "stage": stage})


def silence_prompt(message: str) -> ServerEvent:
    return ServerEvent(type=ServerEventType.SILENCE_PROMPT, data={"message": message})


def assessment_started() -> ServerEvent:
    return ServerEvent(type=ServerEventType.ASSESSMENT_STARTED)


def assessment_completed(session_id: str) -> ServerEvent:
    return ServerEvent(type=ServerEventType.ASSESSMENT_COMPLETED, data={"session_id": session_id})

File: frontend/src/types/websocket.ts
/**
 * TypeScript types matching the backend WebSocket protocol exactly.
 *
 * Source of truth: backend `app/schemas/websocket.py`.
 */

// ── Client-to-server events ───────────────────────────────────────────────

export type ClientEventType =
  | "session.start"
  | "audio.chunk"
  | "audio.end"
  | "candidate.text"
  | "session.end"
  | "ping";

export interface ClientEvent<T = unknown> {
  type: ClientEventType;
  data?: T;
}

export interface CandidateTextData {
  text: string;
}

// ── Server-to-client events ───────────────────────────────────────────────

export type InterviewerState = "speaking" | "listening" | "thinking";

export type ServerEventType =
  | "session.ready"
  | "interviewer.transcript"
  | "interviewer.state"
  | "candidate.transcript"
  | "interviewer.response"
  | "audio.chunk"
  | "audio.end"
  | "silence.prompt"
  | "assessment.started"
  | "assessment.completed"
  | "session.completed"
  | "error"
  | "pong";

export interface ServerEvent<T = Record<string, unknown>> {
  type: ServerEventType;
  data?: T;
}

// ── Typed event payloads ──────────────────────────────────────────────────

export interface SessionReadyData {
  session_id: string;
}

export interface InterviewerTranscriptData {
  text: string;
}

export interface InterviewerStateData {
  state: InterviewerState;
}

export interface CandidateTranscriptData {
  text: string;
}

export interface InterviewerResponseData {
  text: string;
  stage: string;
}

export interface SilencePromptData {
  message: string;
}

export interface SessionCompletedData {
  session_id: string;
}

export interface AssessmentCompletedData {
  session_id: string;
}

export interface ErrorData {
  code: string;
  message: string;
}

// ── Event type → payload mapping ──────────────────────────────────────────

export interface ServerEventMap {
  "session.ready": SessionReadyData;
  "interviewer.transcript": InterviewerTranscriptData;
  "interviewer.state": InterviewerStateData;
  "candidate.transcript": CandidateTranscriptData;
  "interviewer.response": InterviewerResponseData;
  "silence.prompt": SilencePromptData;
  "assessment.started": undefined;
  "assessment.completed": AssessmentCompletedData;
  "session.completed": SessionCompletedData;
  "error": ErrorData;
  "pong": undefined;
  "audio.chunk": undefined;
  "audio.end": undefined;
}

// ── Error codes ───────────────────────────────────────────────────────────

export type WsErrorCode =
  | "SESSION_NOT_FOUND"
  | "SESSION_ALREADY_CONNECTED"
  | "SESSION_ALREADY_COMPLETED"
  | "INVALID_MESSAGE"
  | "AUDIO_TOO_SHORT"
  | "TRANSCRIPTION_FAILED"
  | "TTS_FAILED"
  | "ASSESSMENT_FAILED"
  | "SESSION_NOT_READY";